File: apps/android/app/src/main/java/com/tonezen/app/playback/DownloadQueueState.kt
package com.tonezen.app.playback

import com.tonezen.app.domain.downloads.DownloadResumePolicy

enum class DownloadQueueItemStatus {
    QUEUED,
    DOWNLOADING,
    COMPLETED,
    CANCELLED,
    FAILED,
    PAUSED_OFFLINE,
}

data class DownloadQueueItem(
    val bookId: String,
    val trackId: String,
    val title: String,
    val subtitle: String?,
    val contentType: String,
    val status: DownloadQueueItemStatus,
    val progress: Float?,
    val batchId: String?,
    val enqueuedAt: Long,
    val completedAt: Long?,
)

data class DownloadQueueState(
    val queuedItems: List<DownloadQueueItem> = emptyList(),
    val completedHistory: List<DownloadQueueItem> = emptyList(),
    val activeBookId: String? = null,
    val activeTrackId: String? = null,
    val activeProgress: Float? = null,
    val bulkDownloaded: Int = 0,
    val bulkTotal: Int = 0,
    val activeBatchId: String? = null,
    val pausedForNetwork: Boolean = false,
) {
    val isActive: Boolean
        get() = queuedItems.any {
            it.status == DownloadQueueItemStatus.QUEUED ||
                it.status == DownloadQueueItemStatus.DOWNLOADING ||
                it.status == DownloadQueueItemStatus.PAUSED_OFFLINE
        } || activeTrackId != null

    val isTrackDownloading: Boolean
        get() = activeTrackId != null && activeProgress != null

    val isBulkDownloading: Boolean
        get() = bulkTotal > 0 && bulkDownloaded < bulkTotal

    fun progressForTrack(trackId: String): Float? =
        if (activeTrackId == trackId) activeProgress else null

    fun isTrackQueued(trackId: String): Boolean =
        queuedItems.any {
            it.trackId == trackId &&
                (it.status == DownloadQueueItemStatus.QUEUED || it.status == DownloadQueueItemStatus.PAUSED_OFFLINE)
        }

    val bulkProgress: Float?
        get() {
            if (bulkTotal <= 0) return null
            if (activeTrackId != null && activeProgress != null) {
                return ((bulkDownloaded + activeProgress) / bulkTotal).coerceIn(0f, 1f)
            }
            return (bulkDownloaded.toFloat() / bulkTotal).coerceIn(0f, 1f)
        }
}

fun DownloadQueueState.trimHistory(): DownloadQueueState {
    val limit = DownloadResumePolicy.COMPLETED_HISTORY_LIMIT
    if (completedHistory.size <= limit) return this
    return copy(completedHistory = completedHistory.takeLast(limit))
}

File: apps/desktop/src/core/downloads/downloadQueueState.ts
import { COMPLETED_HISTORY_LIMIT } from "@core/downloads/downloadResumePolicy.js";
import {
  bulkProgressFraction,
  emptyMusicDownloadState,
  isBulkDownloading,
  isMusicDownloadActive,
  isTrackDownloading,
  progressForTrack,
  type MusicDownloadState,
} from "@core/downloads/musicDownloadState.js";
import type { DownloadPriority } from "@core/downloads/downloadQueuePolicy.js";

export type DownloadQueueItemStatus =
  | "QUEUED"
  | "DOWNLOADING"
  | "COMPLETED"
  | "CANCELLED"
  | "FAILED"
  | "PAUSED_OFFLINE";

export type DownloadAwaitResult = "COMPLETED" | "CANCELLED" | "FAILED" | "OFFLINE";

export interface EnqueueDownloadRequest {
  bookId: string;
  trackId: string;
  priority: DownloadPriority;
  batchId?: string | null;
  title: string;
  subtitle?: string | null;
  contentType: string;
  enqueuedAt?: number;
}

export interface DownloadQueueItem {
  bookId: string;
  trackId: string;
  title: string;
  subtitle: string | null;
  contentType: string;
  status: DownloadQueueItemStatus;
  progress: number | null;
  batchId: string | null;
  enqueuedAt: number;
  completedAt: number | null;
}

export interface DownloadQueueState extends MusicDownloadState {
  queuedItems: DownloadQueueItem[];
  completedHistory: DownloadQueueItem[];
  activeBookId: string | null;
  activeBatchId: string | null;
  pausedForNetwork: boolean;
}

export const emptyDownloadQueueState = (): DownloadQueueState => ({
  ...emptyMusicDownloadState(),
  queuedItems: [],
  completedHistory: [],
  activeBookId: null,
  activeBatchId: null,
  pausedForNetwork: false,
});

export function isDownloadQueueActive(state: DownloadQueueState): boolean {
  return (
    state.queuedItems.some(
      (item) =>
        item.status === "QUEUED" ||
        item.status === "DOWNLOADING" ||
        item.status === "PAUSED_OFFLINE",
    ) || state.activeTrackId != null
  );
}

export function isTrackQueued(state: DownloadQueueState, trackId: string): boolean {
  return state.queuedItems.some(
    (item) =>
      item.trackId === trackId &&
      (item.status === "QUEUED" || item.status === "PAUSED_OFFLINE"),
  );
}

export function trimCompletedHistory(state: DownloadQueueState): DownloadQueueState {
  if (state.completedHistory.length <= COMPLETED_HISTORY_LIMIT) return state;
  return {
    ...state,
    completedHistory: state.completedHistory.slice(-COMPLETED_HISTORY_LIMIT),
  };
}

export {
  bulkProgressFraction,
  isBulkDownloading,
  isMusicDownloadActive,
  isTrackDownloading,
  progressForTrack,
};
